### PR TITLE
Hook up the details page to data

### PR DESCRIPTION
**Pull request by kumar303**
Fixes https://github.com/mozilla/addons-frontend/issues/883

### CODE REVIEW

**Comment by kumar303** [comment]
@muffinresearch I bumbled through this one, let me know how it looks. I have not manually tested the install button since I'm not sure what `about:config` settings need tweaking.

File: src/amo/components/AddonDetail.js [hunk 1]
```diff
@@ -1,23 +1,37 @@
...
+      slug: PropTypes.string.isRequired,
+    }),
   }
 
   render() {
-    const { i18n } = this.props;
+    const { i18n, addon } = this.props;
+    const authorList = [];
+    for (const author of addon.authors) {
+      authorList.push(<a key={author.url} href={author.url}>{author.name}</a>);
```
Comment (muffinresearch) [inline src/amo/components/AddonDetail.js]: This feels like it should be a ul since its a list of something. You could then add the comma with generated content.

Comment (muffinresearch) [inline src/amo/components/AddonDetail.js]: Ah now I see it's in an h1, in which case it can't be a ul. Ignore the previous comment.

File: src/amo/components/AddonDetail.js [hunk 2]
```diff
@@ -1,23 +1,37 @@
...
+      authorList.push(<span key={`${author.url}-comma`}>, </span>);
```
Comment (muffinresearch) [inline src/amo/components/AddonDetail.js]: Do you need the span for the comma?

File: src/amo/components/AddonDetail.js [hunk 3]
```diff
@@ -54,17 +67,9 @@ export class AddonDetail extends React.Component {
...
-          iusto nemore, te legere antiopam sit. Suas simul ad usu, ex putent
-          timeam fierent eum. Dicam equidem cum cu. Vel ea vidit timeam.</p>
-
-          <p>Eu nam dicant oportere, et per habeo euismod denique, te appetere
-          temporibus mea. Ad solum reprehendunt vis, sea eros accusata senserit
-          an, eam utinam theophrastus in. Debet consul vis ex. Mei an iusto
-          delicatissimi, ut timeam electram maiestatis nam, te petentium
-          intellegebat ius. Ei legere everti.</p>
+          <div dangerouslySetInnerHTML={sanitizeHTML(nl2br(addon.description),
+                                                     ['b', 'br', 'blockquote', 'cite'])}>
```
Comment (muffinresearch) [inline src/amo/components/AddonDetail.js]: Does this match the list of what we allow at point of entry in amo?

File: src/amo/containers/DetailPage.js [hunk 4]
```diff
@@ -1,13 +1,39 @@
...
+
+function mapStateToProps(state, ownProps) {
+  const { slug } = ownProps.params;
+  return {
+    addon: state.addons[slug],
+    slug,
+  };
+}
+
+export default compose(
```
Comment (muffinresearch) [inline src/amo/containers/DetailPage.js]: This is definitely nicer 👍

File: src/core/utils.js [hunk 5]
```diff
@@ -61,3 +66,48 @@ export function getClientApp(userAgentString) {
 export function isValidClientApp(value, { _config = config } = {}) {
   return _config.get('validClientApplications').includes(value);
 }
+
+purify.addHook('afterSanitizeAttributes', (node) => {
```
Comment (muffinresearch) [inline src/core/utils.js]: This might be better living in `core/purify`?

File: src/core/utils.js [hunk 6]
```diff
@@ -61,3 +66,48 @@ export function getClientApp(userAgentString) {
...
+  // Set all elements owning target to target=_blank
+  // and add rel="noreferrer".
+  if ('target' in node) {
+    node.setAttribute('target', '_blank');
+    node.setAttribute('rel', 'noreferrer');
+  }
+});
+
+export function sanitizeHTML(text, allowTags = []) {
```
Comment (muffinresearch) [inline src/core/utils.js]: I know this was pre-existing, but allowedTags might be an improvement.

File: src/core/utils.js [hunk 7]
```diff
@@ -61,3 +66,48 @@ export function getClientApp(userAgentString) {
...
+  // TODO: Accept tags to allow and run through dom-purify.
+  return {
+    __html: purify.sanitize(text, { ALLOWED_TAGS: allowTags }),
+  };
+}
+
+// Convert new lines to HTML breaks.
+export function nl2br(text) {
+  return text.replace(/\n/g, '<br />');
```
Comment (muffinresearch) [inline src/core/utils.js]: What about other forms of newlines from windows etc?

File: src/amo/components/AddonDetail.js (hunk 2 above)
Comment (kumar303) [inline src/amo/components/AddonDetail.js]: React lists seem to want a `key` attribute on every array item. I might have also hit an error when this was a plain string, let me try again...

Comment (kumar303) [inline src/amo/components/AddonDetail.js]: I removed the span and it seems to be fine now.

File: src/amo/components/AddonDetail.js (hunk 3 above)
Comment (kumar303) [inline src/amo/components/AddonDetail.js]: Good call. I updated it to match olympia.

File: src/amo/components/AddonDetail.js [hunk 8]
```diff
@@ -28,13 +58,13 @@ export class AddonDetail extends React.Component {
             <LikeButton />
           </div>
           <div className="title">
-            <h1>Placeholder Add-on Title
-              <span className="author">by <a href="#">AwesomeAddons</a></span></h1>
-            <InstallButton slug="placeholder" />
+            <h1>
+              {addon.name} <span className="author">{i18n.gettext('by')} {authorList}</span>
```
Comment (kumar303) [inline src/amo/components/AddonDetail.js]: @muffinresearch I almost forgot to localize this. Is it ok to isolate the `by` like this? It seems like creating a longer format string would be complicated because of the span tag.

File: src/core/utils.js (hunk 7 above)
Comment (kumar303) [inline src/core/utils.js]: it looks like [clean_nl()](https://github.com/mozilla/addons-server/blob/master/src/olympia/amo/utils.py#L428) takes care of them.

Comment (kumar303) [inline src/core/utils.js]: actually, bah, that function doesn't do what it says it does

Comment (kumar303) [inline src/core/utils.js]: thanks, fixed

**Comment by kumar303** [comment]
@muffinresearch this is ready for another look

File: src/amo/components/AddonDetail.js (hunk 8 above)
Comment (muffinresearch) [inline src/amo/components/AddonDetail.js]: Good catch!

If we only translate `by` the downside would be if a locale needed to change the word order for some reason it couldn't.

In the past we've done things like this to be able to expose the full context for translation allthought for disco pane the string was built on the server.

```jsx
{i18n.sprintf(i18n.gettext('%(addonName)s %(startSpan)sby %(authorList)s%(endSpan)s'), { addonName, authorList, startSpan, endSpan })}
``` 

If passed through domurify if the tags are reversed via the translation you will still get a sane output albeit the markup won't be wrapping the things it should, but you will at least get valid HTML.

e.g:

```html
</span>foo<span>
```

becomes:

```html
foo<span></span>
```

**Comment by muffinresearch** [comment]
> I have not manually tested the install button since I'm not sure what about:config settings need tweaking. 

Yeah it won't magically work since the APIs aren't going to be exposed to non-disco pane hosts. 
We'll need to start a conversation about opening this up to amo and the implications around it.

File: tests/client/core/test_utils.js [hunk 9]
```diff
@@ -207,3 +214,120 @@ describe('isValidClientApp', () => {
...
+  it('converts \r\n to <br/>', () => {
+    assert.equal(nl2br('\r\n'), '<br />');
+  });
+
+  it('converts \n\n to <br/>', () => {
+    assert.equal(nl2br('\n\n'), '<br /><br />');
+  });
+
+  it('converts \r\n\r\n to <br/>', () => {
+    assert.equal(nl2br('\r\n\r\n'), '<br /><br />');
```
Comment (muffinresearch) [inline tests/client/core/test_utils.js]: Should normalize multiple sets of line breaks to a single `<br />` like the description string suggests?

**Comment by muffinresearch** [comment]
Gave this a spin and it looks good.

r+wc

**Comment by kumar303** [comment]
@muffinresearch if you have a mo' could you check that the new title localization is sane?

File: src/amo/components/AddonDetail.js [hunk 10]
```diff
@@ -39,15 +39,18 @@ class AddonDetail extends React.Component {
...
-      authorList.push(<a key={author.url} href={author.url}>{author.name}</a>);
-      authorList.push(', ');
-    }
-    if (authorList.length) {
-      // Remove the trailing comma.
-      authorList.pop();
-    }
+
+    const authorList = addon.authors.map(
+      (author) => `<a href=${author.url}>${author.name}</a>`);
```
Comment (muffinresearch) [inline src/amo/components/AddonDetail.js]: Missing quotes on the attr.

Comment (kumar303) [inline src/amo/components/AddonDetail.js]: d'oh. Refactoring casualty. Thanks.
